files/default/neutron-ha-tool.py

```diff
@@ -615,63 +614,72 @@ def migrate_l3_routers_from_agent(qclient, agent, targets, agent_picker,
     return (migrations, errors)
 
 
-def migrate_router_safely(qclient, noop, router_id, agent, target,
+def migrate_router_safely(qclient, noop, router, agent, target,
                           wait_for_router=True, delete_namespace=False):
     if noop:
         LOG.info("Would try to migrate router=%s from agent=%s "
-                 "to agent=%s", router_id, agent['id'], target['id'])
+                 "to agent=%s", router['id'], agent['id'], target['id'])
         return True
 
     try:
-        migrate_router(qclient, router_id, agent, target,
+        migrate_router(qclient, router, agent, target,
                        wait_for_router, delete_namespace)
         return True
     except:
         LOG.exception("Failed to migrate router=%s from agent=%s "
-                      "to agent=%s", router_id, agent['id'], target['id'])
+                      "to agent=%s", router['id'], agent['id'], target['id'])
         return False
 
 
-def migrate_router(qclient, router_id, agent, target,
+def migrate_router(qclient, router, agent, target,
                    wait_for_router=True, delete_namespace=False):
     """
     Returns nothing, and raises exceptions on errors.
 
     :param qclient: A neutronclient
-    :param router_id: The id of the router to migrate
+    :param router: The router to migrate
     :param agent_id: The id of the l3 agent to migrate from
     :param target_id: The id of the l3 agent to migrate to
     """
 
     LOG.info("Migrating router=%s from agent=%s to agent=%s",
-             router_id, agent['id'], target['id'])
+             router['id'], agent['id'], target['id'])
 
     # N.B. The neutron API will return "success" even when there is a
     # subsequent failure during the add or remove process so we must check to
     # ensure the router has been added or removed
 
     # Remove the router from the original agent
-    qclient.remove_router_from_l3_agent(agent['id'], router_id)
+    qclient.remove_router_from_l3_agent(agent['id'], router['id'])
     LOG.debug("Removed router from agent=%s" % agent['id'])
 
-    # ensure it is removed or log an error
-    if router_id in list_routers_on_l3_agent(qclient, agent['id']):
-        raise RuntimeError("Failed to remove router_id=%s from agent_id=%s" %
-                           (router_id, agent['id']))
+    # ensure it is removed
+    if router in list_routers_on_l3_agent(qclient, agent['id']):
+        if router['distributed']:
```

Comment:
To avoid confusion on neutron releases that don't suffer from bsc#1016943 (i.e. mitaka and newer) could we please check if the router is still present before trying to delete it a second time?